### PR TITLE
[three] Add three.js types to Definitely Typed

### PR DESCRIPTION
**Pull request by joshuaellis**
THREE.js has removed it's `.d.ts` files, related issue: https://github.com/mrdoob/three.js/pull/21174
In response to this, myself & @Methuselah96 are moving them to the DefinitelyTyped repo. I'm creating a draft so the folks at three can follow it & also so if current maintainers of DT have opinions on the merge that they can be heard early on.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/three/src/loaders/Loader.d.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    resourcePath: string;
+    manager: LoadingManager;
+
+    /**
+     * @default {}
+     */
+    requestHeader: { [header: string]: string };
+
+    /*
+	load(): void;
```
Comment (danger-public) [inline types/three/src/loaders/Loader.d.ts:36]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

**Comment by danger-public** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **types/three/examples/jsm/modifiers/EdgeSplitModifier.d.ts#L7** - The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.
  </td>
    </tr>
  </tbody>
</table>



Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## three ([<kbd>unpkg</kbd>](https://unpkg.com/browse/three@latest/))
was missing the following properties:
1. wrap
1. isGeneratorFunction
1. mark
1. awrap
1. AsyncIterator

<details>
<summary>as well as these 35 other properties...</summary>
<p>async, keys, values, AxisHelper, BinaryTextureLoader, BoundingBoxHelper, CanvasRenderer, ClosedSplineCurve3, CustomToneMapping, DynamicBufferAttribute, EdgesHelper, Face3, Face4, FaceColors, JSONLoader, LensFlare, Math, MeshFaceMaterial, MultiMaterial, NoColors, Particle, ParticleBasicMaterial, ParticleSystem, ParticleSystemMaterial, PointCloud, PointCloudMaterial, SceneUtils, Spline, SplineCurve3, Vertex, VertexColors, WebGLRenderTargetCube, WebGLUtils, WireframeHelper, XHRLoader</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against db5b65bd0995de7931b0d8a34a107df8477aa4b4
</p>

File: types/three/src/core/Object3D.d.ts [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+    setRotationFromEuler(euler: Euler): void;
+
+    setRotationFromMatrix(m: Matrix4): void;
+
+    setRotationFromQuaternion(q: Quaternion): void;
+
+    /**
+     * Rotate an object along an axis in object space. The axis is assumed to be normalized.
+     * @param axis	A normalized vector in object space.
```
Comment (danger-public) [inline types/three/src/core/Object3D.d.ts:229]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

Comment (joshuaellis) [reply types/three/src/core/Object3D.d.ts:229]: @Methuselah96 is there a quick way we can edit the files to do this? 😅

File: types/three/examples/jsm/modifiers/EdgeSplitModifier.d.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { BufferGeometry } from '../../../src/Three';
+
+export class EdgeSplitModifier {
+    constructor();
+
+    /**
+     * @param geometry					The geometry to modify by splitting edges.
```
Comment (danger-public) [inline types/three/examples/jsm/modifiers/EdgeSplitModifier.d.ts:7]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

Comment (joshuaellis) [reply types/three/examples/jsm/modifiers/EdgeSplitModifier.d.ts:7]: We've ran prettier, but this didn't update.

**Comment by joshuaellis** [comment]
good bot.

File: types/physijs/index.d.ts [hunk 4]
```diff
@@ -1,250 +0,0 @@
-// Type definitions for Physijs
```
Comment (amcasey) [inline types/physijs/index.d.ts:1]: How is this deletion related to the main change?

File: types/three/OTHER_FILES.txt [hunk 5]
```diff
@@ -0,0 +1,303 @@
+examples/jsm/animation/AnimationClipCreator.d.ts
```
Comment (amcasey) [inline types/three/OTHER_FILES.txt:1]: What are these?  Why are there 300?

File: types/three/src/Three.Legacy.d.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export {};
```
Comment (amcasey) [inline types/three/src/Three.Legacy.d.ts:1]: What's this?

**Review by amcasey — COMMENTED** [review]
We can't practically review 584 files, but I assume they're mostly copied from already-working files in the underlying library.  I added questions about some things I found unclear.

File: types/three/src/Three.Legacy.d.ts (hunk 6 above)
Comment (joshuaellis) [reply types/three/src/Three.Legacy.d.ts:1]: Copied directly from the original library.

File: types/three/OTHER_FILES.txt (hunk 5 above)
Comment (joshuaellis) [reply types/three/OTHER_FILES.txt:1]: The `three` repo has two sections of its source code - `src` & `examples`. We're currently not able to write enough tests for every file, so @Methuselah96 made the decision to move the examples into `OTHER_FILES` as we still need them bunlded.

File: types/physijs/index.d.ts (hunk 4 above)
Comment (joshuaellis) [reply types/physijs/index.d.ts:1]: This is a question for @Methuselah96

**Comment by joshuaellis** [comment]
Hi @amcasey,

Sure, it's definitely unrealistic. Three is a huge library! I've answered your Q's above, although some of them need the other developer to answer. If you have any more, feel free to ask.

File: types/physijs/index.d.ts (hunk 4 above)
Comment (Methuselah96) [reply types/physijs/index.d.ts:1]: Hey, sorry for the late reply. I just asked a question about this on Discord.

> Is there a way to have subfolders for old versions that that are still in the 0.x.y releases? The use case is three.js. physijs relies on @types/three@^0.93.31, but the current version is 0.125.x. The only way that I can think to deal with it is to remove the types for that package, but that doesn't seem ideal (however the package hasn't been updated in about 6 years and gets about 6 weekly downloads so I'm not sure it's actually getting used).

File: types/three/OTHER_FILES.txt (hunk 5 above)
Comment (Methuselah96) [reply types/three/OTHER_FILES.txt:1]: Yeah, they're not included in the main export of the package, but they can be imported via their path (e.g., `import { AnimationClipCreator } from 'three/examples/jsm/animation/AnimationClipCreator';`).

File: types/physijs/index.d.ts (hunk 4 above)
Comment (Methuselah96) [reply types/physijs/index.d.ts:1]: Talked to @sandersn about this on Discord. Since the package is practically dead, he said that it was fine to go ahead and delete the types and add it to `notNeededPackages.json` (for the purposes of getting the build to pass).

Comment (Methuselah96) [reply types/physijs/index.d.ts:1]: Oh, wait it's actually accurate. `physijs` added a definition file in April: https://github.com/loperd/physijs.

Comment (peterblazejewicz) [reply types/physijs/index.d.ts:1]: Shouldn't `physisjs` be updated to depends on updated `threejs` version: So when it's ships to NPM (with dts changes) should fix that problem with client consumption of types:
https://github.com/loperd/physijs/blob/master/package.json#L7

File: types/three/src/Three.Legacy.d.ts (hunk 6 above)
Comment (peterblazejewicz) [reply types/three/src/Three.Legacy.d.ts:1]: TL;DR; it should not be empty, but IMO that's OK for now

@joshuaellis
Should it contain empty types resassembling current library types? If I understand it was added to provide feedback, warnings or errors when using legacy objects:
https://github.com/mrdoob/three.js/blob/945ecfd7ea53aebf1d43e8b29ca6e48e084609be/src/Three.Legacy.js

File: types/physijs/index.d.ts (hunk 4 above)
Comment (Methuselah96) [reply types/physijs/index.d.ts:1]: I'm not sure I follow. When the maintainers of `physijs` upgrade to a higher version of `three` that no longer has the types, they'll have to add a dependency on `@types/three`, but since they're still depending on a version of `three` that contains types it should be fine for now.

What exactly are you proposing?

File: types/three/src/Three.Legacy.d.ts (hunk 6 above)
Comment (joshuaellis) [reply types/three/src/Three.Legacy.d.ts:1]: @peterblazejewicz, the way the repo is composed, those are extending classes/functions elsewhere in the repo, for example, `Uniform` the `dynamic` property is written in the Legacy file, but the type is written [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50930/files#diff-ee11c4160e1f7b605d53dec6f6575c2a7fa90e31ab41052a720bb511704e7f38R15).

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
thx folks!

**Review by amcasey — APPROVED** [review]

**Comment by joshuaellis** [comment]
Ready to merge
